controllers/clusterwidenetworkpolicy_controller.go

```diff
@@ -19,54 +17,193 @@ package controllers
 func (r *ClusterwideNetworkPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	var clusterNP firewallv1.ClusterwideNetworkPolicy
-	if err := r.Get(ctx, req.NamespacedName, &clusterNP); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	log := r.log.WithValues("cwnp", req.Name)
+
+	var cwnps firewallv1.ClusterwideNetworkPolicyList
+	if err := r.List(ctx, &cwnps, client.InNamespace(firewallv1.ClusterwideNetworkPolicyNamespace)); err != nil {
+		return done, err
+	}
+
+	return r.reconcileRules(ctx, log, cwnps)
+}
+
+func (r *ClusterwideNetworkPolicyReconciler) reconcileRules(ctx context.Context, log logr.Logger, cwnps firewallv1.ClusterwideNetworkPolicyList) (ctrl.Result, error) {
+	var f firewallv1.Firewall
+	nn := types.NamespacedName{
+		Name:      firewallName,
+		Namespace: firewallv1.ClusterwideNetworkPolicyNamespace,
 	}
+	if err := r.Get(ctx, nn, &f); err != nil {
+		if apierrors.IsNotFound(err) {
+			return done, err
+		}
 
-	// if network policy does not belong to the namespace where clusterwide network policies are stored:
-	// update status with error message
-	if req.Namespace != firewallv1.ClusterwideNetworkPolicyNamespace {
-		r.recorder.Event(&clusterNP, corev1.EventTypeWarning, "Unapplicable", fmt.Sprintf("cluster wide network policies must be defined in namespace %s otherwise they won't take effect", firewallv1.ClusterwideNetworkPolicyNamespace))
-		return ctrl.Result{}, nil
+		return done, client.IgnoreNotFound(err)
 	}
 
-	err := clusterNP.Spec.Validate()
+	// Set CWNP requeue interval
+	if interval, err := time.ParseDuration(f.Spec.Interval); err == nil {
+		r.interval = interval
+	}
+
+	if err := r.manageDNSProxy(f, cwnps, log); err != nil {
+		return done, err
+	}
+
+	var services corev1.ServiceList
+	if err := r.List(ctx, &services); err != nil {
+		return done, err
+	}
+	nftablesFirewall := r.createFirewall(&cwnps, &services, f.Spec, r.cache, log)
+	updated, err := nftablesFirewall.Reconcile()
 	if err != nil {
-		r.recorder.Event(&clusterNP, corev1.EventTypeWarning, "Unapplicable", fmt.Sprintf("cluster wide network policy is not valid: %v", err))
-		return ctrl.Result{}, nil
+		return done, err
 	}
 
-	return ctrl.Result{}, nil
+	if updated {
+		for _, i := range cwnps.Items {
+			o := i
+			if err := r.Update(ctx, &o); err != nil {
+				return done, err
+			}
+		}
+	}
+
+	return done, nil
+}
+
+// manageDNSProxy start DNS proxy if toFQDN rules are present
+// if rules were deleted it will stop running DNS proxy
+func (r *ClusterwideNetworkPolicyReconciler) manageDNSProxy(f firewallv1.Firewall, cwnps firewallv1.ClusterwideNetworkPolicyList, log logr.Logger) (err error) {
+	// Skipping is needed for testing
+	if r.skipDNS {
+		return nil
+	}
+
+	enableDNS := len(cwnps.GetFQDNs()) > 0
+	kb := network.GetUpdatedKnowledgeBase(f)
+	nftablesFirewall := nftables.NewDefaultFirewall()
+	nftablesFirewall.ReconcileNetconfTables(kb, enableDNS)
```

Comment:
It's recreating only `metal-networker` part of nftables: `inet metal`, `inet nat` tables. It's done here because there's rules that need to be added/removed if DNS is switched on/off.

The `firewall-controller` part of nftables is rendered by `firewall.Reconcile` method in `reconcileRules`.

Reply:
Okay, it's coming clearer to me. One more question: Why can't the nftablesFirewall from the caller function be used?